src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -360,74 +410,90 @@ public JsonPropertyInfo GetProperty(ReadOnlySpan<byte> propertyName, ref ReadSta
+            if (length > 7)
+            {
+                key = MemoryMarshal.Read<ulong>(propertyName);
+
+                // Clear the high byte so it can hold the length.
+                key &= 0x00FFFFFFFFFFFFFF;
+
+                // Use a length of 8. Any length > 7 (PropertyNameKeyLength) would work since
+                // the comparison logic tests for equality against the full contents instead of just
+                // the key if the property name is 8 or more characters.
+                key |= (ulong) 8 << (7 * BitsInByte);
+            }
+            else if (length > 3)
             {
                 key = MemoryMarshal.Read<uint>(propertyName);
-                if (length > 4)
+
+                if (length == 7)
+                {
+                    key |= (ulong) propertyName[4] << (4 * BitsInByte)
+                        | (ulong) propertyName[5] << (5 * BitsInByte)
+                        | (ulong) propertyName[6] << (6 * BitsInByte)
+                        | (ulong) 7 << (7 * BitsInByte);
+                }
+                else if (length == 6)
                 {
-                    key |= (ulong)propertyName[4] << 32;
+                    key |= (ulong) propertyName[4] << (4 * BitsInByte)
+                        | (ulong) propertyName[5] << (5 * BitsInByte)
+                        | (ulong) 6 << (7 * BitsInByte);
                 }
-                if (length > 5)
+                else if (length == 5)
+                {
+                    key |= (ulong) propertyName[4] << (4 * BitsInByte)
+                        | (ulong) 5 << (7 * BitsInByte);
+                }
+                else
                 {
-                    key |= (ulong)propertyName[5] << 40;
+                    key |= (ulong) 4 << (7 * BitsInByte);
                 }
             }
             else if (length > 1)
             {
                 key = MemoryMarshal.Read<ushort>(propertyName);
-                if (length > 2)
+
+                if (length == 3)
                 {
-                    key |= (ulong)propertyName[2] << 16;
+                    key |= (ulong) propertyName[2] << (2 * BitsInByte)
+                        | (ulong) 3 << (7 * BitsInByte);
+                }
+                else
+                {
+                    key |= (ulong) 2 << (7 * BitsInByte);
                 }
             }
             else if (length == 1)
             {
-                key = propertyName[0];
+                key = propertyName[0]
+                    | (ulong) 1 << (7 * BitsInByte);
             }
             else
             {
                 // An empty name is valid.
                 key = 0;
             }
 
-            // Embed the propertyName length in the last two bytes.
-            key |= (ulong)propertyName.Length << 48;
+            // Verify key contains the embedded bytes as expected.
+            Debug.Assert(
+                length < 1 || propertyName[0] == (key & ((ulong)0xFF << 8 * 0)) >> 8 * 0 &&
```

Comment:
I think && bind the conditions stronger than ||.

So, these conditions are like:
```C#
(length < 1) ||
(propertyName[0] == (key & ((ulong)0xFF << 8 * 0)) >> 8 * 0 && length < 2) ||
(propertyName[1] == (key & ((ulong)0xFF << 8 * 1)) >> 8 * 1) && length < 3) ||
...
```

And are not like, which seem like the intention:
```C#
(length < 1 || propertyName[0] == (key & ((ulong)0xFF << 8 * 0)) >> 8 * 0) &&
(length < 2 || propertyName[1] == (key & ((ulong)0xFF << 8 * 1)) >> 8 * 1) &&
...
```

What is the correct intention here? And if the intention was to bind the ||s together, why is this `Debug.Assert` not failing?

```C#
byte[] someArray = new byte[2];
someArray[0] = 1;
someArray[1] = 1;

// No brackets - && wins
bool result = someArray.Length < 1 || someArray[0] == 0 && 
              someArray.Length < 2 || someArray[1] == 1;
Console.WriteLine(result); // True

// Add brackets to bind || - answer changes
result = (someArray.Length < 1 || someArray[0] == 0) && 
              (someArray.Length < 2 || someArray[1] == 1);
Console.WriteLine(result); // False

// Add brackets explicitly bind &&
result = someArray.Length < 1 || (someArray[0] == 0 && 
              someArray.Length < 2) || someArray[1] == 1;
Console.WriteLine(result); /// True
```